### PR TITLE
[RFC] Implement custom parser API

### PR DESCRIPTION
**Pull request by azz**
Similar to #1710, this PR allows external AST modification, however it does so by re-using the `--parser` option.

If `parser` is a function, that function will be used in place of a built-in function. The parser may fall back to one of prettier's built-in parsers via the second argument.

If `parser` is a string that's not one of the built-in parsers, it will be resolved as a Node.js module.

## API

```ts
prettier.format(code: string, options: {
  parser: string | (text: string, parsers: PrettierParsers) => AST
}): string;
```

## Example

```js
const output = prettier.format("foo ( )", {
  parser(text, { babylon }) {
    expect(typeof babylon).toEqual("function");
    const ast = babylon(text);
    ast.program.body[0].expression.callee.name = "bar";
    return ast;
  }
});
expect(output).toEqual("bar();\n");
```

### CODE REVIEW

**Comment by azz** [comment]
Added some docs.

https://github.com/azz/prettier/tree/feat/parser-module#custom-parser-api

File: README.md [hunk 1]
```diff
@@ -273,25 +296,29 @@ prettier.formatWithCursor(" 1", { cursorOffset: 2 });
 // -> { formatted: '1;\n', cursorOffset: 1 }
 ```
 
-### Options
+#### Custom Parser API
 
-Prettier ships with a handful of customizable format options, usable in both the CLI and API.
+If you need to make modifications to the AST (such as codemods), or you want to provide an alternate parser, you can do so setting the `parser` option to a function. The function signature of the parser function is:
```
Comment (existentialism) [inline README.md]: > you can do so _by_ setting

**Comment by vjeux** [comment]
Looks good to me. Could you double check that the release version still works (run build.sh and read the instructions to run the tests).

If it does, feel free to merge it

File: scripts/build/build.sh [hunk 2]
```diff
@@ -72,10 +78,7 @@ cp package.json dist/
 echo 'Done!'
 echo;
 echo 'How to test against dist:'
-echo '  1) Open tests_config/run_spec.js'
-echo '  2) add `dist/` to the require'
-echo '  3) yarn test'
-echo "  4) Don't forget to revert tests_config/run_spec.js"
+echo '  1) yarn test --prod'
```
Comment (azz) [inline scripts/build/build.sh]: Much better 😉

**Pull request merged**